Rx/v2/src/rxcpp/rx-coroutine.hpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved. See License.txt in the project root for license information.
+
+#pragma once
+
+/*! \file rx-coroutine.hpp
+
+    \brief The proposal to add couroutines to the standard adds `co_await`, `for co_await`, `co_yield` and `co_return`. This file adds `begin(observable<>)` & `end(observable<>)` which enables `for co_await` to work with the `observable<>` type.
+
+    for co_await (auto c : interval(seconds(1), observe_on_event_loop()) | take(3)) {
+        printf("%d\n", c);
+    }
+
+*/
+
+#if !defined(RXCPP_RX_COROUTINE_HPP)
+#define RXCPP_RX_COROUTINE_HPP
+
+#include "rx-includes.hpp"
+
+#ifdef _RESUMABLE_FUNCTIONS_SUPPORTED
+
+#include <rxcpp/operators/rx-finally.hpp>
+
+#include <experimental/resumable>
+
+namespace rxcpp {
+namespace coroutine {
+
+using namespace std;
+using namespace std::chrono;
+using namespace std::experimental;
+
+template<typename Source>
+struct co_observable_iterator;
+
+template<typename Source>
+struct co_observable_iterator_state
+{
+    using value_type = typename Source::value_type;
+
+    explicit co_observable_iterator_state(const Source& o) : o(o) {}
+
+    coroutine_handle<> caller{};
+    composite_subscription lifetime{};
+    const value_type* value{nullptr};
+    exception_ptr error{nullptr};
+    Source o;
+};
+
+template<typename Source>
+struct co_observable_inc_awaiter
+{
+    bool await_ready() {
+        return false;
+    }
+
+    bool await_suspend(const coroutine_handle<>& handle) {
```

Comment:
I would drop const& so that the parameter type is just coroutine_handle<> . The coroutine_handle is just a fancy template around a void* with no destructor, so passing it by value is cheaper.